Jetpack E2E: add coverage for the newsletter subscription & removal flow.

worldomonation opened the pull request:
Related to https://github.com/Automattic/wp-calypso/issues/80730.

## Proposed Changes

This PR adds E2E coverage to the newsletter subscribe/removal steps.

Key flows:
- subscribe to a site via the Subscribe block.
- verify the subscription.
- as the site owner, remove the subscriber.

## Testing Instructions

Ensure the following build configurations are passing:
  - [x] Calypso E2E (desktop)
  - [x] Jetpack E2E Simple (mobile)
  - [x] Jetpack E2E Simple (desktop)
  - [x] Jetpack E2E AT (desktop)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
packages/calypso-e2e/src/rest-api-client.ts
- add new methods to get and delete subscribers

packages/calypso-e2e/src/secrets/secrets-manager.ts
- add new secret under mailosaur

worldomonation committed:
packages/calypso-e2e/src/secrets/encrypted.enc
- update encrypted file

worldomonation committed:
packages/calypso-e2e/src/lib/pages/published/published-post-page.ts
- add method to subscribe to the site.

packages/calypso-e2e/src/lib/pages/subscribers-page.ts
- new POM to interact with the Users > Subscribers page.

packages/calypso-e2e/src/lib/pages/subscription-management-page.ts
- new POM to interact with the subscription management endpoint available without an account.

worldomonation committed:
test/e2e/specs/users/newsletter__subscribe-remove.ts
- new E2E to walk through the free newsletter subscribe and removal steps.

worldomonation committed:
test/e2e/specs/users/newsletter__subscribe-remove.ts
- add new group

worldomonation committed:
packages/calypso-e2e/src/data-helper.ts
- deprecate the getTestEmailAddress method.

packages/calypso-e2e/src/email-client.ts
- add new method getTestEmailAddress which uses the mailosaur method.

test/e2e/specs/users/newsletter__subscribe-remove.ts
- adjust E2E

worldomonation committed:
packages/calypso-e2e/src/lib/pages/published/published-post-page.ts
- update and add try/catch to the subscribe modal

dpasque commented on packages/calypso-e2e/src/email-client.ts:
```diff
@@ -17,6 +17,14 @@ export class EmailClient {
 	}
 
+	/**
+	 *
```
**Nit**: FYI -- looks like you might be missing the rest of the doc block here

worldomonation replied:
I knew I missed something!

dpasque commented on packages/calypso-e2e/src/email-client.ts:
```diff
@@ -86,6 +94,36 @@ export class EmailClient {
+		}
+
+		const links = message.html.links as Link[];
+		const regexp = new RegExp( key, 'i' );
```
**Suggestion (non-blocking)**: Making a RegEx out of an unvalidated string seems potentially risky, as it won't escape RegEx special characters by default.

Is the goal to just find a link that has text matching the provided key? Could we just do a substring search on `link.text`?

worldomonation replied:
> Is the goal to just find a link that has text matching the provided key?

Correct! An instance of a `Link` would look something like this

```typescript
{
    "link": <url>,
    "text": "Continue"
}
```

A key in this case would be "Continue". 

> Could we just do a substring search on link.text?

I think that would work, yes.

dpasque commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -420,6 +420,11 @@ function setupBrowserProxyTrap( browser: Browser ): Browser {
 					} );
 
+					// Abort ad loading on ad-enabled Jetpack AT sites.
+					page.route( /pubmine\.com/, ( route ) => {
```
Oh, nice idea!

dpasque commented on packages/calypso-e2e/src/lib/pages/published/published-post-page.ts:
```diff
@@ -83,6 +83,39 @@ export class PublishedPostPage {
+		// shown.
+		try {
+			// Site has paid plans.
+			await iframe
+				.getByRole( 'link', { name: 'Free - Get a glimpse of the newsletter' } )
+				.waitFor( { timeout: 6 * 1000 } );
```
**Question**: I think I'm a little bit confused on this line -- is the user supposed to click this link? Or is the goal to do extra validation?

worldomonation replied:
Hopefully I can clarify!

The idea here is to handle two situations:
- site has a paid newsletter subscription plan already set up (eg. $5/month).
- site does not yet have subscription set up.

The sequence of dialog popup UI differs slightly depending on which scenario. It can be visualized somewhat like this:

```
entrypoint: user has filled out the subscription block's input, and clicked on the "Subscribe" button in the published post.

site has newsletter plan ------> dialog presents choice of payment option and the free trial link -----> user opts for free trial -----> dialog states a confirmation email has been sent.

site has no newsletter plan ------> dialog states a confirmation email has been sent.
```
To visualize, take look at the following screenshots:

![image](https://github.com/Automattic/wp-calypso/assets/6549265/04106ebe-1e65-4d36-b4fc-928db6ac0835)
![image](https://github.com/Automattic/wp-calypso/assets/6549265/ccb5649f-a020-460c-a36b-8dfe4f8915fa)

Because we can't know beforehand whether a newsletter subscription plan is defined for a site (from the perspective of a user), I opted to use the try/catch to handle the unknown.

dpasque commented on packages/calypso-e2e/src/lib/pages/subscription-management-page.ts:
```diff
@@ -0,0 +1,40 @@
+		this.anchor = this.page.getByRole( 'main' );
+	}
+
+	/**
+	 * Validates the user is subscribed to a specific site, either by
+	 * the site name or the site URL.
+	 *
+	 * @param param0 Keyed object parameter.
+	 * @param {string} [param0.siteName] Name of the site.
+	 * @param {string} [param0.siteURL] URL of the site.
+	 * @throws {Error} If no params are defined.
+	 */
+	async validateSubscribedSiteBy( {
+		siteName,
+		siteURL,
+	}: { siteName?: string; siteURL?: string } = {} ) {
+		if ( ! siteName && ! siteURL ) {
+			throw new Error( 'Must define either of `siteName` or `siteURL`.' );
+		}
+		// Yes, this OR statement actually works!
+		await this.anchor.getByRole( 'link', { name: siteName || siteURL } ).waitFor();
+	}
```
**Suggestion (non-blocking)**: I like the descriptiveness of this signature, but it feels a little weird and error prone because there are two optional options, but one is technically required. What do you think of making it take a single string parameter called "identifier"? Or having two overloaded functions with different parameter names but that then have the same shared implementation?

worldomonation replied:
Hmm, you're right. I guess we can put the onus of supplying a valid identifier to the calling method.

dpasque commented on test/e2e/specs/users/newsletter__subscribe-remove.ts:
```diff
@@ -0,0 +1,136 @@
+
+		afterAll( async function () {
+			try {
+				// Clean up the subscriber and the post.
+				await restAPIClient.deleteSubscriber(
+					testAccount.credentials.testSites?.primary.id as number,
+					testEmail
+				);
```
This is perfectly fine here, but I also think there is a potential job we can use on the backend as part of our clean-up to unlink all the users/subscribers on a test site -- let me know if you want us to start using that here! 😄

worldomonation replied:
That sounds cool, I'm guessing you mean adding to the list of tasks the empty blog function is expected to perform, correct?

worldomonation replied:
I'm going to keep this in here for the time being and we can discuss whether this needs to be removed when you return from AFK next week.

dpasque commented on packages/calypso-e2e/src/lib/pages/subscribers-page.ts:
```diff
@@ -0,0 +1,100 @@
+	 *
+	 * @param {NewSubscribersList} subscribers List of emails to add, up to 3.
+	 */
+	async addSubscribers( subscribers: NewSubscribersList ) {
```
**Question (non-blocking)** Looking at the test -- do we need this function? I'm fine leaving it in fwiw, was just curious. 🙂

worldomonation replied:
We don't, for now - this spec was initially scoped out purely as an interaction by the site owner, before shifting to the current form. I can remove it for now.

dpasque reviewed (approved):
Nice, this looks like it was a tricky one to put together! Thanks for taking it on! 😄

worldomonation committed:
packages/calypso-e2e/src/email-client.ts
- fill out missing JSDoc

worldomonation committed:
packages/calypso-e2e/src/email-client.ts
- replace regex with substring search.

worldomonation committed:
test/e2e/specs/users/newsletter__subscribe-remove.ts
packages/calypso-e2e/src/lib/pages/subscription-management-page.ts
- update method signature.
- revise method name.

worldomonation committed:
packages/calypso-e2e/src/lib/pages/subscribers-page.ts
- remove unnecessary code.

packages/calypso-e2e/src/jest-playwright-config/environment.ts
- remove route abort, which is implemented elsewhere.

worldomonation committed:
packages/calypso-e2e/src/lib/pages/published/published-post-page.ts
- use OR locator

worldomonation committed:
packages/calypso-e2e/src/email-client.ts
- implement new method to get link by the link text.

packages/calypso-e2e/src/jest-playwright-config/environment.ts
- add pre-emptive abort to kill connections to ad servers
